src/main/ruby/truffleruby/core/integer.rb

```diff
@@ -59,9 +60,57 @@ def **(o)
     redo_coerced :**, o
   end
 
-  def [](index)
+  def [](index, len = undefined)
+    if index.kind_of?(Range)
+      handle_range(index)
+    else
+      handle_aref(index, len)
+    end
+  end
+
+  private def handle_aref(index, len)
     index = Primitive.rb_to_int(index)
-    index < 0 ? 0 : (self >> index) & 1
+    if Primitive.undefined?(len)
+      index < 0 ? 0 : (self >> index) & 1
+    else
+      num = self >> index
+      mask = (1 << len) - 1
+
+      num & mask
+    end
+  end
+
+  private def handle_range(range)
+    validate_range(range)
+
+    if !Primitive.nil?(range.begin) && !Primitive.nil?(range.end)
+      len = range.end - range.begin
+      num = self >> range.begin
+      mask = mask(range, len)
+
+      range.end < range.begin ? num : num & mask
+    elsif Primitive.nil? range.end
+
+      return self >> range.begin
+    else
+      result = self & mask(range, range.end)
+      raise ArgumentError, 'The beginless range for Integer#[] results in infinity' if result != 0 && range.end >= 0
+
+      0
+    end
+  end
+
+  private def validate_range(index)
+    raise FloatDomainError , 'Infinity' if index.begin == Float::INFINITY || index.end == Float::INFINITY
+    raise FloatDomainError , '-Infinity' if index.begin == -Float::INFINITY || index.end == -Float::INFINITY
+  end
+
+  private def mask(range, idx)
```

Comment:
Could you move this logic in the two callers, and e.g. do `len += 1 unless range.exclude_end?` so then it's just `(1 << len) - 1`?